src/traversals/bfs.jl

```diff
@@ -13,65 +13,52 @@ type BreadthFirst <: SimpleGraphVisitAlgorithm
 end
 
 function breadth_first_visit_impl!(
-    graph::SimpleGraph,   # the graph
-    queue::Vector{Int},                  # an (initialized) queue that stores the active vertices
-    colormap::Vector{Int},          # an (initialized) color-map to indicate status of vertices
-    visitor::SimpleGraphVisitor)  # the visitor
+    graph::SimpleGraph,                 # the graph
+    queue::Vector{Int},                 # an (initialized) queue that stores the active vertices
+    vertexcolormap,                           # an (initialized) color-map to indicate status of vertices (0=unseen, 1=seen, 2=closed)
+    edgecolormap,                           # an (initialized) color-map to indicate status of edges
+    visitor::SimpleGraphVisitor)        # the visitor
 
     while !isempty(queue)
         u = shift!(queue)
         open_vertex!(visitor, u)
 
         for v in out_neighbors(graph, u)
-            v_color::Int = colormap[v]
-            # TODO: Incorporate edge colors to BFS
-            if !(examine_neighbor!(visitor, u, v, v_color, -1))
-                return
-            end
+            v_color = get(vertexcolormap, v, 0)
+            v_edge = Edge(u,v)
+            e_color = get(edgecolormap, v_edge, 0)
+            examine_neighbor!(visitor, u, v, v_color, e_color) || return
+
+            edgecolormap[v_edge] = 1
 
             if v_color == 0
-                colormap[v] = 1
+                vertexcolormap[v] = 1
                 discover_vertex!(visitor, v) || return
                 push!(queue, v)
             end
         end
 
-        colormap[u] = 2
         close_vertex!(visitor, u)
+        vertexcolormap[u] = 2
     end
-    nothing
-end
-
-function traverse_graph(
-    graph::SimpleGraph,
-    alg::BreadthFirst,
-    s::Int,
-    visitor::SimpleGraphVisitor;
-    colormap = zeros(Int, nv(graph)),
-    que = Vector{Int}())
-
-    colormap[s] = 1
-    discover_vertex!(visitor, s) || return
-    push!(que, s)
-
-    breadth_first_visit_impl!(graph, que, colormap, visitor)
 end
 
-function traverse_graph(
+function traverse_graph!(
     graph::SimpleGraph,
     alg::BreadthFirst,
-    sources::AbstractVector{Int},
+    source,
     visitor::SimpleGraphVisitor;
-    colormap = zeros(Int, nv(graph)),
-    que = Vector{Int}())
+    vertexcolormap = Dict{Int, Int}(),
+    edgecolormap = DummyEdgeMap(),
```

Comment:
...and this is where we're going to run into trouble when we insist on typing all variables.